design-documents/graph-ql/mutation-error-design.md

```diff
@@ -0,0 +1,154 @@
+
+If we move our known error states for various mutations into the schema design itself, we'll get a few benefits:
+
+1. Versioning of errors
+2. Discoverability of error states for UI devs
+
+### Example: Fixing our `addItemToCart` mutation
+
+In the previous example of `addItemToCart`, we had the following schema:
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): Cart
+}
+```
+
+Let's design some known error states directly into the API
+
+```graphql
+type Mutation {
+    addItemToCart(sku: String!, quantity: Float): AddItemToCartOutput
+}
+
+type AddItemToCartOutput {
+    # can still get the entire state of the cart post-mutation
+    cart: Cart
+    # can directly access errors that a shopper should be notified about
+    addItemUserErrors: [AddItemUserError!]!
+}
+
+type AddItemUserError {
+    # if a UI has their own text for message, they can just
+    # not use this field, but serves as a descriptive default
+    message: string!
+    type: AddItemUserErrorType!
+}
+
+# This enum is just an example. Non-exhaustive, and of course
+# not all errors we'd want a specific type for
+enum AddItemUserErrorType {
+    OUT_OF_STOCK
+    MAX_QTY_FOR_USER
+    NOT_AVAILABLE
+}
+```
+
+With this new schema, we get the following query and response:
+
+```graphql
+mutation {
+    addItemToCart(sku: "abc", quantity: 1) {
+        cart {
+            items {
+                # select item fields
+            }
+        }
+
+        addItemUserErrors {
+            message
+            type
+        }
+    }
+}
+```
+
+```js
+{
+    "data": {
+        "cart": {
+            // cart data here
+        },
+        "addItemUserErrors": [{
+            "message": "Item 'cool-hat' not added to cart (Out of Stock)",
+            "type": "OUT_OF_STOCK"
+        }]
+    }
+}
+```
+## Should we still use the `errors` key at all?
+
+Yes! _But_, we should consider the `errors` key to be of similar use to the `catch` clause when using `try/catch` in a programming language.
+
+The `errors` key should be for _exceptional_ circumstances, not for describing well-known states in an ecommerce app. It's unlikely you'd write this application code on the backend:
+
+```javascript
+try {
+    var product = addItemToCart('sku');
+    return product;
+} catch (error) {
+    if (error.code === 'OUT_OF_STOCK') {
+        //
+    } else if (error.code === 'MAX_QTY_FOR_USER') {
+        //
+    } else {
+        // something else failed, maybe the DB connection?
+    }
+}
+```
+
+Instead, you would likely have your `addItemToCart` function return something that describes the various possible results of the operation. 
+
+## Open Questions
+
+1. Should we be consistent with the field name that represents these first-class errors? `userErrors` vs something like `addToCartUserErrors`
+2. Should there be a basic interface that mutations should implement to enforce the pattern of returning a list of user errors?
```

Comment:
Great document!  Would it make sense to incorporate union types rather than relying on each response to include an error portion?  This allows for common errors to be standardized across multiple calls in a uniform way, such as extending your example with multiple methods such as "addToCart", "addAllToCart", "putWishlistItemInCart", and so on.  This could be handled using interfaces as previously suggested but union type force the caller to acknowledge the error case.  It also provides for cleaner separation between happy path and failure path objects. 
 For example:

```graphql
type Mutation {
    addItemToCart(sku: String!, quantity: Float): AddItemToCartOutput
    putWishlistItemInCart(wishlistId: String!, sku: String): WishlistItemMovedToCartOutput
}

type CartUpdatedOutput {
    cart: Cart
}

type WishlistUpdatedOutput {
    cart: Cart
    wishlist: Wishlist
}

type CartUpdateError {
    message: string!
    reasons: [CartUpdateErrorReason!]!
}

union AddItemToCartOutput = CartUpdatedOutput | CartUpdateError
union WishlistItemMovedToCartOutput = WishlistUpdatedOutput | CartUpdateError

# and so on...
```

Reply:
Thanks for the feedback @billkalter! I'm a big fan of this approach, and it's similar to [the approach that artsy took](https://artsy.github.io/blog/2018/10/19/where-art-thou-my-error/) for their API. If I was building a new GraphQL API today, this is likely the approach I'd go with.

The main reason I opted not to suggest this approach is that we already have a fair number of mutations today, and we'd have to deprecate each one and introduce a new mutation if we wanted to change the result type to a union.

Because we already have requirements around so-called "output objects" for all mutations (missed on a few very old ones), I thought it would make sense for us to piggyback on this and avoid the breaking changes.

Thoughts?

<details>
<summary>Click to see Mutation type from Magento core</summary>

```graphql
    type Mutation {
        addBundleProductsToCart(
            input: AddBundleProductsToCartInput
        ): AddBundleProductsToCartOutput
        addConfigurableProductsToCart(
            input: AddConfigurableProductsToCartInput
        ): AddConfigurableProductsToCartOutput
        addDownloadableProductsToCart(
            input: AddDownloadableProductsToCartInput
        ): AddDownloadableProductsToCartOutput
        addSimpleProductsToCart(
            input: AddSimpleProductsToCartInput
        ): AddSimpleProductsToCartOutput
        addVirtualProductsToCart(
            input: AddVirtualProductsToCartInput
        ): AddVirtualProductsToCartOutput
        applyCouponToCart(
            input: ApplyCouponToCartInput
        ): ApplyCouponToCartOutput

        """
        Apply a pre-defined gift card code to the specified cart.
        """
        applyGiftCardToCart(
            input: ApplyGiftCardToCartInput
        ): ApplyGiftCardToCartOutput

        """
        Apply store credit to the specified cart.
        """
        applyStoreCreditToCart(
            input: ApplyStoreCreditToCartInput!
        ): ApplyStoreCreditToCartOutput

        """
        Changes the password for the logged-in customer
        """
        changeCustomerPassword(
            currentPassword: String!
            newPassword: String!
        ): Customer

        """
        Creates Client Token for Braintree Javascript SDK initialization.
        """
        createBraintreeClientToken: String!

        """
        Create customer account
        """
        createCustomer(input: CustomerInput!): CustomerOutput

        """
        Create customer address
        """
        createCustomerAddress(input: CustomerAddressInput!): CustomerAddress

        """
        Creates an empty shopping cart for a guest or logged in user
        """
        createEmptyCart(input: createEmptyCartInput): String

        """
        Initiates a transaction and receives a token. Use this mutation for Payflow Pro and Payments Pro payment methods
        """
        createPayflowProToken(
            input: PayflowProTokenInput!
        ): CreatePayflowProTokenOutput

        """
        Initiates an Express Checkout transaction and receives a token. Use this
        mutation for Express Checkout and Payments Standard payment methods.
        """
        createPaypalExpressToken(
            input: PaypalExpressTokenInput!
        ): PaypalExpressTokenOutput

        """
        Delete customer address
        """
        deleteCustomerAddress(id: Int!): Boolean

        """
        Delete a customer payment token
        """
        deletePaymentToken(public_hash: String!): DeletePaymentTokenOutput

        """
        Retrieve the customer token
        """
        generateCustomerToken(email: String!, password: String!): CustomerToken

        """
        Handles payment response and saves payment in Quote. Use this mutations for Payflow Pro and Payments Pro payment methods.
        """
        handlePayflowProResponse(
            input: PayflowProResponseInput!
        ): PayflowProResponseOutput

        """
        Merges the source cart into the destination cart
        """
        mergeCarts(source_cart_id: String!, destination_cart_id: String!): Cart!
        placeOrder(input: PlaceOrderInput): PlaceOrderOutput

        """
        Redeem gift card for store credit.
        """
        redeemGiftCardBalanceAsStoreCredit(
            input: GiftCardAccountInput!
        ): GiftCardAccount
        removeCouponFromCart(
            input: RemoveCouponFromCartInput
        ): RemoveCouponFromCartOutput
        removeGiftCardFromCart(
            input: RemoveGiftCardFromCartInput
        ): RemoveGiftCardFromCartOutput
        removeItemFromCart(
            input: RemoveItemFromCartInput
        ): RemoveItemFromCartOutput

        """
        Remove applied store credit from the specified cart.
        """
        removeStoreCreditFromCart(
            input: RemoveStoreCreditFromCartInput!
        ): RemoveStoreCreditFromCartOutput

        """
        Adds all products from a customer's previous order to the cart.
        """
        reorderItems(orderNumber: String!): ReorderItemsOutput

        """
        Revoke the customer token
        """
        revokeCustomerToken: RevokeCustomerTokenOutput

        """
        Recommends Product by Sending Single/Multiple Email
        """
        sendEmailToFriend(
            input: SendEmailToFriendInput
        ): SendEmailToFriendOutput
        setBillingAddressOnCart(
            input: SetBillingAddressOnCartInput
        ): SetBillingAddressOnCartOutput
        setGuestEmailOnCart(
            input: SetGuestEmailOnCartInput
        ): SetGuestEmailOnCartOutput
        setPaymentMethodAndPlaceOrder(
            input: SetPaymentMethodAndPlaceOrderInput
        ): PlaceOrderOutput
            @deprecated(
                reason: "Should use setPaymentMethodOnCart and placeOrder mutations in single request."
            )
        setPaymentMethodOnCart(
            input: SetPaymentMethodOnCartInput
        ): SetPaymentMethodOnCartOutput
        setShippingAddressesOnCart(
            input: SetShippingAddressesOnCartInput
        ): SetShippingAddressesOnCartOutput
        setShippingMethodsOnCart(
            input: SetShippingMethodsOnCartInput
        ): SetShippingMethodsOnCartOutput
        updateCartItems(input: UpdateCartItemsInput): UpdateCartItemsOutput

        """
        Update the customer's personal information
        """
        updateCustomer(input: CustomerInput!): CustomerOutput

        """
        Update customer address
        """
        updateCustomerAddress(
            id: Int!
            input: CustomerAddressInput
        ): CustomerAddress
    }
```
</details>